### PR TITLE
Hypre mat type

### PR DESCRIPTION
**Pull request by fdkong**
Enable hypre mat type

### CODE REVIEW

File: test/tests/kernels/2d_diffusion/tests [hunk 1]
```diff
@@ -70,4 +70,14 @@
     issues = '#11760'
     design = 'FEProblem.md'
   [../]
+
+  [./test_hypre_matrix]
```
Comment (permcody) [inline test/tests/kernels/2d_diffusion/tests]: This is not a good place to add this test. Why not add this to a "solvers" or "preconditioners" folder?

File: framework/include/problems/FEProblemBase.h [hunk 2]
```diff
@@ -1861,6 +1863,8 @@ class FEProblemBase : public SubProblem, public Restartable
   /// requested an AD material property or whether any suppier has declared an AD material property
   bool _using_ad_mat_props;
 
+  bool _hypre_matrix;
```
Comment (permcody) [inline framework/include/problems/FEProblemBase.h]: I would prefer _reuse_hypre_matrix for a Boolean. As it stands, it's a little confusing and you might think it's a pointer to the matrix.

File: framework/src/executioners/Executioner.C [hunk 3]
```diff
@@ -133,6 +133,11 @@ validParams<Executioner>()
                         false,
                         "Should XFEM update the mesh at the beginning of the timestep");
 
+  params.addParam<bool>("hypre_matrix",
```
Comment (permcody) [inline framework/src/executioners/Executioner.C]: How about "reuse_hypre_matrix"?

File: framework/src/problems/EigenProblem.C [hunk 4]
```diff
@@ -247,3 +247,13 @@ EigenProblem::isNonlinearEigenvalueSolver()
          solverParams()._eigen_solve_type == Moose::EST_MONOLITH_NEWTON ||
          solverParams()._eigen_solve_type == Moose::EST_MF_MONOLITH_NEWTON;
 }
+
+void
+EigenProblem::init()
+{
+#ifdef LIBMESH_HAVE_PETSC
+  Moose::PetscSupport::petscSetOptions(*this); // Make sure the PETSc options are setup for this app
```
Comment (permcody) [inline framework/src/problems/EigenProblem.C]: Why is this not being done in the base class?

File: framework/src/utils/PetscSupport.C [hunk 5]
```diff
@@ -968,6 +968,26 @@ colorAdjacencyMatrix(PetscScalar * adjacency_matrix,
   ISColoringDestroy(&iscoloring);
 }
 
+bool
+detectHypreFromOptions()
+{
+  char pc_type[256];
+  PetscBool flg;
```
Comment (permcody) [inline framework/src/utils/PetscSupport.C]: It's a dang good thing you didn't type that "a". You saved soooo much space there!

File: framework/src/executioners/Executioner.C [hunk 6]
```diff
@@ -133,6 +133,11 @@ validParams<Executioner>()
...
+                        false,
```
Comment (permcody) [inline framework/src/executioners/Executioner.C]: Why not default to true?

File: test/tests/kernels/2d_diffusion/tests [hunk 7]
```diff
@@ -70,4 +70,14 @@
...
+    type = 'Exodiff'
+    input = '2d_diffusion_test.i'
```
Comment (permcody) [inline test/tests/kernels/2d_diffusion/tests]: Also, is there any way to tell that we are saving memory? How are we _actually_ testing this?

**Review by permcody — CHANGES_REQUESTED** [review]

**Comment by fdkong** [comment]
Open open open and open.

File: framework/src/utils/PetscSupport.C (hunk 5 above)
Comment (dschwen) [reply framework/src/utils/PetscSupport.C]: image

**Pull request closed without merging**